src/components/FlagsAutocomplete/FlagsAutocomplete.tsx

```diff
@@ -0,0 +1,227 @@
+  AutocompleteProps<
+    unknown,
+    boolean | undefined,
+    boolean | undefined,
+    boolean | undefined
+  >
+> & {
+  anchorEl: HTMLElement | null
+  continents?: MuiTelInputContinent[]
+  excludedCountries?: MuiTelInputCountry[]
+  flagSize?: FlagSize
+  isoCode: MuiTelInputCountry | null
+  langOfCountryName?: string
+  onClose: () => void
+  onlyCountries?: MuiTelInputCountry[]
+  onSelectCountry: (isoCode: MuiTelInputCountry) => void
+  preferredCountries?: MuiTelInputCountry[]
+}
+
+export type MuiTelAutocompleteOption = {
+  countryCode: MuiTelInputCountry
+  callingCode: string
+  displayName: string
+}
+
+const FlagsAutocomplete = (props: FlagsAutocompleteProps) => {
+  const {
+    anchorEl,
+    onSelectCountry,
+    excludedCountries,
+    onlyCountries,
+    langOfCountryName,
+    continents,
+    preferredCountries,
+    className,
+    flagSize,
+    onClose
+  } = props
+
+  // Idem for the translations
+  const displayNames = React.useMemo(() => {
+    return getDisplayNames(langOfCountryName)
+  }, [langOfCountryName])
+
+  const ISO_CODES_SORTED = sortAlphabeticallyCountryCodes(
+    ISO_CODES,
+    displayNames
+  )
+
+  const countriesFiltered = filterCountries(ISO_CODES_SORTED, {
+    onlyCountries,
+    excludedCountries,
+    continents,
+    preferredCountries
+  })
+
+  const countriesFilteredOptions: MuiTelAutocompleteOption[] =
+    countriesFiltered.map((countryCode) => {
+      return {
+        countryCode,
+        callingCode: COUNTRIES[countryCode]?.[0] as string,
+        displayName: displayNames.of(countryCode) ?? countryCode
+      }
+    })
+
+  const filterOptions = (
+    options: MuiTelAutocompleteOption[],
+    { inputValue }: FilterOptionsState<MuiTelAutocompleteOption>
+  ) => {
+    if (inputValue === '') {
+      return options
+    }
+
+    return matchSorter(options, inputValue, {
+      keys: ['callingCode', 'countryCode', 'displayName']
+    })
+  }
+
+  const getOptionLabel = (option: string | MuiTelAutocompleteOption) => {
+    if (typeof option === 'string') {
+      return ''
+    }
+
+    return option.countryCode
+  }
+
+  return (
+    <Styled.FlagsAutocompletePopper
+      anchorEl={anchorEl}
+      className={`MuiTelInput-FlagsAutocomplete-Popover ${className || ''}`}
+      id="muitelinput-flagsautocomplete"
+      open={Boolean(anchorEl)}
+      placement="bottom-start"
+    >
+      <ClickAwayListener onClickAway={onClose}>
+        <Autocomplete
+          autoHighlight
+          filterOptions={filterOptions}
+          freeSolo
+          getOptionLabel={getOptionLabel}
+          onChange={(event, newValue, reason) => {
+            if (
+              event.type === 'keydown' &&
+              (event as React.KeyboardEvent).key === 'Backspace' &&
+              reason === 'removeOption'
+            ) {
+              return
+            }
+
+            if (newValue !== null) {
+              if (typeof newValue !== 'string') {
```

Comment:
what is happening here ?

Reply:
Added comments to the code to explain what's going on here.